datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -1160,6 +1162,24 @@ pub(crate) fn from_substrait_type(dt: &substrait::proto::Type) -> Result<DataTyp
                     "Unsupported Substrait type variation {v} of type {s_kind:?}"
                 ),
             },
+            r#type::Kind::UserDefined(u) => {
+                match u.type_reference {
+                    INTERVAL_YEAR_MONTH_TYPE_REF => {
```

Comment:
I may be totally wrong, but are you sure this is how `type_reference` is supposed to work? I'd kind of expect the type_reference to map to an extension / MappingType::ExtensionType, kinda like [function_reference](https://github.com/apache/datafusion/blob/2e6beb434d45693bc8c7ba636b0407c26c04fc37/datafusion/substrait/src/logical_plan/consumer.rs#L747).

Reply:
I am not at all sure how this is supposed to work -- when I was reviewing this PR I think I concluded it followed the existing patterns.

Reply:
There are two references

https://github.com/apache/datafusion/pull/10646/files#diff-d1c5f4c37ac8286d2045acb61bee17382179469557132eb02844413b260ae41bR1441-R1442

Reply:
I see, actually I think both references (type_variation_reference and type_reference) have the same problem - I hadn't realized it affects the type_variation_reference too. Now, if each system defines the type/variation references as consts, then plans will look compatible, but may have different meaning (nothing tells another Substrait producer/consumer that type_variation_reference = 1 on a list means a large list, for example).

I believe both should be defined as SimpleExtension files, following the [schema](https://github.com/substrait-io/substrait/blob/4f5b4ac4d473c9f03f30f86eca080073d99ef1c7/text/simple_extensions_schema.yaml#L17) (like here https://github.com/apache/arrow/blob/main/format/substrait/extension_types.yaml) rather than as constants (kinda what https://github.com/apache/datafusion/blob/09054263df1d8de06b6e77ab0cbd99027bb7ceb6/datafusion/substrait/src/variation_const.rs#L21 already says 😅). And then (I think) in the plan itself, there'd be instances of the SimpleExtensionDeclaration messages (https://github.com/substrait-io/substrait/blob/4f5b4ac4d473c9f03f30f86eca080073d99ef1c7/proto/substrait/extensions/extensions.proto#L39), and the type_reference and type_variation_reference would link to the type_anchor and type_variation_anchor, rather than to the hard-coded constants. 

That way one can (in theory, at least) teach another Substrait producer/consumer about the DataFusion extensions and keep plans compatible, or at least the systems will recognize that the plans are incompatible as it refers to extension URIs that the other system doesn't know about.

Does that make sense? I'm not 100% sure of anything I'm saying here, so I may as be understanding something wrong.

Reply:
>I believe both should be defined as SimpleExtension files

Yes, that's right. The last (I'm aware of, at least) big piece we are missing in substrait is those *.yaml spec for all extended things and related URL settings. At present, all the things are defined in the document.

From the substrait [website](substrait.io), we'll need a yaml parsing component to support extensions from other systems as well, if we are going to implement the ability to consume plans from external systems.

Reply:
going to file a tracking issue for tasks related to substrait support

Reply:
Updated at https://github.com/apache/datafusion/issues/8149

Reply:
Ah cool, so it was the plan all along :) Sg!

Reply:
Document PR: https://github.com/apache/datafusion/pull/10719